### ISSUE TITLE
7장 앙상블 학습과 랜덤 포레스트

### ISSUE THREAD

**Issue by kangchanguk** [kangchanguk/study_ml#7]
### 앙상블 학습
- 일련의 예측기(분류나 회귀 모델)에서 예측을 수집하여 가장 좋은 모델 하나보다 더 좋은 예측을 얻는 학습법

#### 랜덤 포레스트
- 각기 다른 서브 셋을 만들어 일련의 결정트리 분류기를 훈련, 예측 시 모든 개별 트리의 예측을 구하고 가장 많은 선택을 받은 클래스를 예측으로 선택

### 투표 기반 분류기
- 각 분류기의 예측을 모아서 가장 많이 선택된 클래스를 예측하는 것(직접 투표 분류기)
- 다수결 투표 분류기가 앙상블에 포함된 개별 분류기 중 가장 뛰어난 것보다 정확도가 높을 경우가 많다.
- 다만 모든 분류기가 독립적이고 상관관계가 없어야 가능

EX) 동전 앞/뒷면 예측 분류기
1) 앞면 51%, 뒷면 49%가 나오는 동전이 존재
2) 수학적으로(이항 분포의 확률 질향 함수)를 통해 앞면이 다수가 될 확률은 75%에 가까움
3) 큰 수의 법칙에 의해 앞면이 다수가 될 확률은 더욱 더 증가함
4) 이렇게 (1000번 던져서 앞면이 나오는지 뒷면나오는지를 학습시킨 모델)이 여러개가 있다면 75%의 정확도를 기대해 볼 수 있음

```
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import VotingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.datasets import make_moons
from sklearn.model_selection import train_test_split

log_clf = LogisticRegression()
rnd_clf = RandomForestClassifier()
svm_clf = SVC()

x, y  = make_moons(n_samples=500, noise=0.15, random_state=42)
x_train, x_val, y_train, y_val = train_test_split(x,y, random_state=42)
voting_clf = VotingClassifier(
    estimators = [("lr", log_clf), ("rf", rnd_clf), ("svc", svm_clf)],
    voting = "hard"
)

voting_clf.fit(x_train, y_train)

from sklearn.metrics import accuracy_score

for clf in(log_clf, rnd_clf, svm_clf, voting_clf):
    clf.fit(x_train, y_train)
    y_pred = clf.predict(x_val)
    print(clf.__class__.__name__, accuracy_score(y_val, y_pred))
```

- 간접 투표: 모든 분류기가 클래스의 확률을 예측할 수 있다면 개별 분류기의 예측을 평균 내어 확률이 가장 높은 클래스를 예측하는 방식
- 간접 투표 방식을 사용하기 위해서는 voting 옵션을 "soft"로 변경

## 배깅과 페이스팅
- 다양한 분류기를 만드는 한 가지 다른 방법은 각기 다른 훈련 알고리즘을 사용하는 것 외에 같은 알고리즘을 사용하는 대신 훈련 세트의 서브셋을 무작위로 구성하여 분류기를 각기 다르게 학습시키는 것

- 훈련 세트에서 중복을 허용하여 샘플링하는 방식을 **배깅**, 중복을 허용하지 않고 샘플링하는 방식을 **페이스팅**이라고 함 

![image](https://user-images.githubusercontent.com/26477881/127764000-21515b68-f2a4-426b-95ce-5430c04939a3.png)

- 모든 예측기가 훈련이 끝나면 앙상블은 모든 예측기의 예측을 모아서 새로운 샘플에 대한 예측을 만듬
   - 분류 일때는 통계적 최빈값(가장 많은 예측 결과)를 선택
   - 회귀 일때는 평균값을 활용
- 개별 예측기는 원본 훈련 세트로 훈련 시킨 것보다 훨씬 편향되어 있지만 수집함수를 통과하면서 편향과 분산이 모두 감소함
- 일반적으로 앙상블의 결과는 원본 데이터셋으로 하나의 예측기를 훈련시킬 때와 비교해 편향은 비슷하지만 분산은 줄어듬

### 사이킷런의 배경과 페이스팅
- BaggingClassifier은 결정 트리 분류기 500개의 앙상블을 훈련시키는 코드
- 각 분류기는 훈련세트에서 중복을 허용하여 무작위로 선택된 100개의 샘플로 훈련됨
- 앙상블의 예측은 결정 트리 하나의 예측보다 일반화가 훨씬 잘된 것을 확인 가능
- 부트스트래핑은 서브셋에 다양성을 증가시키므로 배깅이 페이스팅보다 편향이 높음
- 다양성을 추가한다는 것은 예측기들의 상관관계를 줄이는 것이므로 앙상블의 분산을 감소시킴

```
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier

bag_clf = BaggingClassifier(
    DecisionTreeClassifier(), n_estimators = 500,
    max_samples=100, bootstrap=True, n_jobs=-1)

bag_clf.fit(x_train, y_train)
y_pred = bag_clf.predict(x_val)
```

### oob 평가
- 배깅을 사용하면 어떤 샘플은 한 예측기를 위해 여러 번 샘플링 되고 어떤 것은 전혀 선택되지 않을 수 있음'
- BaggingClassifier의 경우 기본적으로 중복을 허용하기 때문에 전체 샘플의 약 63%만 선택되고 나머지 37%는 선택되지 않음
- 이 선택되지 않은 37%의 샘플을 oob(out of bag)
- oob_score=True 설정을 사용하면 oob 평가를 자동으로 실행한다.
- oob_score와 accuracy_score는 거의 동일

### 랜덤 포레스트

- 배깅 방법을 적용한 결정트리의 앙상블
- max_samples를 훈련세트의 크기로 지정
- RandomForestClassifier를 사용할 수 있음

```
from sklearn.ensemble import RandomForestClassifier

rnd_clf = RandomForestClassifier(n_estimators = 500, max_leaf_nodes = 16, n_jobs = -1)
rnd_clf.fit(x_train, y_train)

y_pred_rf=rnd_clf.predict(x_test)
```

- 랜덤 포레스트 알고리즘의 경우 트리의 노드를 분할할 때 전체 특성 중에서 최선의 특성을 찾는 대신 무작위로 선택한 특성 후보중에서 최적의 특성을 찾는 식으로 무작위성을 더욱 주입함
- 다양성을 높임으로서 편향을 손해보는 대신에 분산을 낮춰 전체적으로 더 훌륭한 모델을 만들어 냄

#### 엑스트라 트리
- 랜덤 프로세트에서 트리를 만들 때 각 노드는 무작위로 특성의 서브셋을 만들어 분할에 사용함
- 트리를 더욱 무작위하게 만들기 위해 최적의 임계값을 찾는 대신 후보 특성을 사용하여 무작위로 분할한 다음 그중에서 최상의 분할을 선택함
- 이와 같이 극단적으로 무작위한 트리의 랜덤 포레스트를 익스트림 랜덤 트리 앙상블(엑스트라 트리)라고 함
- 편향이 늘어나지만 분산은 줄어듬
- 모든 노드에서 특성마다 가장 최적의 임계값을 찾는 것이 트리 알고리즘에서 가장 시간이 많이 소모되는 작업이기에 일반적인 랜덤 포레스트보다 빠름
- ExtraTreesClassifier를 활용

#### 특성의 중요도
- 랜덤 포레스트의 또 다른 장점은 특성의 상대적 중요도를 측정하기 쉽다는 것
- 사이킷런은 어떤 특성을 사용한 노드가 평균적으로 불순도를 얼마나 감소시키는지 확인하여 특성의 중요도를 측정합니다
- feature_importances 변수에 저장되어 있음

```
from sklearn.datasets import load_iris

iris = load_iris()
rnd_clf = RandomForestClassifier(n_estimators=500, n_jobs=-1)
rnd_clf.fit(iris["data"], iris["target"])
for name, score in zip(iris["feature_names"], rnd_clf.feature_importances_):
    print(name, score)
```

### 부스팅

- 약한 학습기를 여러개 연결하여 강한 학습기를 만드는 앙상블 방법
- 앞의 모델을 보완하면서 일련의 예측기를 학습시키는 방식
- adaBoost, gradient Boost

### adaboost
1) 첫번째 분류기를 훈련 세트에서 훈련시키고 예측을 만든다.
2) 잘못 분류된 훈련 샘플의 가중치를 상대적으로 높이고 학습
3) 업데이트 된 가중치를 사용하여 훈련 세트에서 훈련하고 다시 예측을 만듬

![image](https://user-images.githubusercontent.com/26477881/127768456-033ad4eb-8936-4a22-beaa-8162336fd694.png)

- 연속된 학습 기법은 경사하강법과 유사한 점이 존재(에이다 부스트는 성능 개선을 위해 앙상블에 예측기를 추가함)

```
from sklearn.ensemble import AdaBoostClassifier

ada_clf = AdaBoostClassifier(
    DecisionTreeClassifier(max_depth=1), n_estimators=200,
    algorithm="SAMME.R", learning_rate=0.5
)
ada_clf.fit(x_train, y_train)
```

### 그레이디언트 부스팅
- 이전까지의 오차를 보정하도록 예측기를 순차적으로 추가합니다.
- 이전 예측기가 만든 잔여 오차에 새로운 예측기를 학습시킴
```
from sklearn.tree import DecisionTreeRegresssor

tree_reg1 = DecisionTreeRegressor(max_depth=2)
tree_reg1.fit(x,y)

y2 = y - tree_reg1.predict(x)
tree_reg2 = DecisionTreeRegressor(max_depth=2)
tree_reg2.fit(x,y2)

y3 = y2 - tree_reg2.predict(X)
tree_reg2 = DecisionTreeRegressor(max_depth=2)
tree_reg2.fit(x, y3)
```

```
from sklearn.ensemble import GradientBoostingRegressor

gbrt = GradientBoostingRegressor(max_depth=2, n_estimators=3, learning_rate=1.0)
gbrt.fit(x,y)
```

- n_estimators: 트리 수를 결정
- learning_rate: 각 트리의 기여 정도를 의미

![image](https://user-images.githubusercontent.com/26477881/127768939-b0368a56-e64d-42d1-baf1-96d59fb30463.png)

- learning_rate에 따라 필요한 트리 갯수가 달라짐
- 트리 갯수가 너무 많아지면 과대 적합이 발생하기에 규제가 필요함
- 대표적인 규제 방법으로는 조기 종료 기법이 있다.

```
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

x_train, x_val, y_train, y_val = train_test_split(x,y)

gbrt = GradientBoostingRegressor(max_depth=2, n_estimators=120)
gbrt.fit(x_train, y_train)

errors= [mean_squared_error(y_val, y_pred) for y_pred in gbrt.staged_predict(x_val)]
bst_n_estimators = np.argmin(errors) + 1

gbrt_best = GradientBoostingRegressor(max_depth=2, n_estimators=bst_n_estimators)
gbrt_best.fit(x_train, y_train)
```
- 120개의 트리로 GBRT 앙상블을 훈련시키고 최적의 트리 수를 찾기 위해 각 훈련 단계에서 검증 오차를 측정한다.
- 마지막에 최적의 트리수를 활용해 새로운 GBRT 앙상블을 훈련시킴

```
gbrt = GradientBoostingRegressor(max_depth=2, warm_start=True)

min_val_error = float("inf")
error_going_up = 0
for n_estimators in range(1,120):
    gbrt.n_estimators = n_estimators
    gbrt.fit(x_train, y_train)
    y_pred = gbrt.predict(x_val)
    val_error = mean_squared_error(y_val, y_pred)
    if val_error < min_val_error:
        min_val_error = val_error
        error_going_up = 0
    else:
        error_going_up +=1
        if error_going_up == 5:
            break
```

### gradient boosting(확률적 그레이디언트 부스팅)

- 각 트리가 훈련할 때 사용할 훈련 샘플의 비율을 지정할 수 있는 subsample 매개변수를 지원
- 부분적인 훈련 샘플로 학습함으로써 편향은 높아지지만 분산은 낮아짐

```
import xgboost

xgb_reg = xgboost.XGBRegressor()
xgb_reg.fit(x_train, y_train)
y_pred = xgb_reg.predict(x_val)

xgb_reg.fit(x_train, y_train, eval_set=[x_val, y_val], early_stopping_rounds=2)
y_pred = xgb_reg.predict(x_val)
```

### 스태킹
- 앙상블에 속한 모든 예측기의 예측을 취합하는 모델
- 마지막 예측기(블렌더)가 최종 예측을 만듬
- 홀드 아웃 세트를 활용

1) 원본 데이터 train_set이 존재
2) 원본 train_data를 3개의 머신러닝 모델이 학습
3) 각 모델 마다 x_test를 넣어서 예측 후 predict를 뽑아냄
4) 3개의 predict를 다시 학습 데이터로 사용
5) 최종 model을 하나 선정해 학습
6) 최종 평가
